v1.1.0 build fails because it cannot find some OpenSSL bindings

Hello,

When trying to build Kryoptic v1.1.0 release tarball with the default features, within a Debian Bookworm (12) Docker container with both `libssl3` and `libssl-dev` installed, I get the following 2 errors:

```
error[E0425]: cannot find value `OSSL_SIGNATURE_PARAM_CONTEXT_STRING` in this scope
   --> src/ossl/eddsa.rs:220:30
    |
220 |                 name_as_char(OSSL_SIGNATURE_PARAM_CONTEXT_STRING),
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `OSSL_SIGNATURE_PARAM_DIGEST_SIZE`
    |
   ::: /kryoptic/target/debug/build/kryoptic-67f0933e799feefa/out/ossl_bindings.rs:602:1
    |
602 | pub const OSSL_SIGNATURE_PARAM_DIGEST_SIZE: &[u8; 12] = b"digest-size\0";
    | ------------------------------------------------------------------------- similarly named constant `OSSL_SIGNATURE_PARAM_DIGEST_SIZE` defined here

error[E0425]: cannot find value `OSSL_SIGNATURE_PARAM_INSTANCE` in this scope
   --> src/ossl/eddsa.rs:244:22
    |
244 |         name_as_char(OSSL_SIGNATURE_PARAM_INSTANCE),
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `OSSL_SIGNATURE_PARAM_DIGEST`
    |
   ::: /kryoptic/target/debug/build/kryoptic-67f0933e799feefa/out/ossl_bindings.rs:597:1
    |
597 | pub const OSSL_SIGNATURE_PARAM_DIGEST: &[u8; 7] = b"digest\0";
    | -------------------------------------------------------------- similarly named constant `OSSL_SIGNATURE_PARAM_DIGEST` defined here
```

This doesn't happen when I build locally on my machine, however.

--- a/.github/workflows/openssl_versions.yml
+++ b/.github/workflows/openssl_versions.yml
@@ -1,0 +1,175 @@
+---
+name: OpenSSL Versions
+
+on:
+  pull_request:
+    branches: ["main"]
+  push:
+    branches: ["main"]
+
+jobs:
+  build:
+    name: Versions
+    runs-on: ubuntu-22.04
+    strategy:
+      fail-fast: false
+      matrix:
+        name: [ossl30, ossl32, ossl35, debian, centos10]
+        include:
+          - name: ossl30
+            build: true
+            version: "3.0"
+            repotype: "dnf"
+            repocachepath: "/var/cache/libdnf5"
+            container: fedora:latest
+          - name: ossl32
+            build: true
+            version: "3.2"
+            repotype: "dnf"
+            repocachepath: "/var/cache/libdnf5"
+            container: fedora:latest
+          - name: ossl35
+            build: true
+            version: "3.5"
+            repotype: "dnf"
+            repocachepath: "/var/cache/libdnf5"
+            container: fedora:latest
+          - name: debian
+            build: false
+            version: "3.5"
+            repotype: "apt"
+            container: debian:testing
+          - name: centos10
+            build: false
+            version: "3.5"
+            repotype: "dnf"
+            repocachepath: "/var/cache/dnf"
+            container: quay.io/centos/centos:stream10
+    container: ${{ matrix.container }}
+    steps:
+      - name: Get Date for DNF cache entry
+        if: ${{ matrix.repotype == 'dnf' }}
+        id: get-date
+        run: |
+          echo "date=$(/bin/date -u "+%Y%V")" >> $GITHUB_OUTPUT
+        shell: bash
+
+      - name: Restore DNF cache
+        if: ${{ matrix.repotype == 'dnf' }}
+        uses: actions/cache/restore@v4
+        id: cache-dnf
+        with:
+          path: ${{ matrix.repocachepath }}
+          key: ${{ matrix.container }}-dnf-${{ steps.get-date.outputs.date }}
+
+      - name: Install Dependencies
+        run: |
+          if [ "${{ matrix.repotype }}" = "dnf" ]; then
+            if [ "${{ matrix.name }}" = "centos" ]; then
+              dnf_opts="--enablerepo=crb"
+            fi
+            dnf -y install git cargo clang-devel
+            if [ "${{ matrix.build }}" = "false" ]; then
+              dnf -y install openssl-devel sqlite-devel
+            else
+              dnf -y install openssl-devel sqlite-devel \
+                'perl(FindBin)' 'perl(lib)' 'perl(File::Compare)' \
+                'perl(File::Copy)' 'perl(bigint)' 'perl(Time::HiRes)' \
+                'perl(IPC::Cmd)' 'perl(Pod::Html)' 'perl(Digest::SHA)' \
+                'perl(Module::Load::Conditional)' 'perl(File::Temp)' \
+                'perl(Test::Harness)' 'perl(Test::More)' 'perl(Math::BigInt)' \
+                zlib-devel sed sqlite-devel
+            fi
+          elif [ "${{ matrix.repotype }}" = "apt" ]; then
+            apt-get -q update
+            apt-get -yq install git clang rustc libsqlite3-dev libssl-dev
+          fi
+
+      - name: DNF cache
+        if: ${{ matrix.repotype == 'dnf' && steps.cache-dnf.outputs.cache-hit != 'true' }}
+        uses: actions/cache/save@v4
+        with:
+          path: ${{ matrix.repocachepath }}
+          key: ${{ matrix.container }}-dnf-${{ steps.get-date.outputs.date }}
+
+
+      - name: Checkout Repository
+        uses: actions/checkout@v4
+
+      - name: Setup OpenSSL ${{ matrix.version }}
+        if: ${{ matrix.build == true }}
+        id: ossl-setup
+        run: |
+          OPENSSL_BRANCH="openssl-${{ matrix.version }}"
+
+          git config --global --add safe.directory /__w/kryoptic
+          cd ..
+          git clone https://github.com/openssl/openssl.git \
+                    --single-branch --branch $OPENSSL_BRANCH openssl
+          cd openssl
+          echo "KRYOPTIC_OPENSSL_SOURCES=$PWD" >> "$GITHUB_ENV"
+          echo "cacheid=${{ runner.os }}-ossl-$(git rev-parse HEAD)" >> "$GITHUB_OUTPUT"
+
+      - name: Restore OpenSSL build if cached
+        if: ${{ matrix.build == true }}
+        uses: actions/cache/restore@v4
+        id: cache
+        with:
+          path: ${{ env.KRYOPTIC_OPENSSL_SOURCES }}
+          key: ${{ steps.ossl-setup.outputs.cacheid }}
+
+      - name: Build OpenSSL
+        if: ${{ matrix.build == true && steps.cache.outputs.cache-hit != 'true' }}
+        run: |
+          pushd ${{ env.KRYOPTIC_OPENSSL_SOURCES }}
+          ./Configure
+          make
+
+      - name: Cache OpenSSL ${{ matrix.version }} build
+        if: ${{ matrix.build == true && steps.cache.outputs.cache-hit != 'true' }}
+        uses: actions/cache/save@v4
+        with:
+          path: ${{ env.KRYOPTIC_OPENSSL_SOURCES }}
+          key: ${{ steps.ossl-setup.outputs.cacheid }}
+
+      - name: Generate lock file
+        run: cargo generate-lockfile
+
+      - name: Cache Rust dependencies
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            target/
+          key: ${{ matrix.container }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Build & test
+        run: |
+          FEATURES="none"
+          OPTS="--no-default-features"
+
+          if [ "${{ matrix.version }}" = "3.0" ]; then
+            FEATURES="minimal"
+          elif [ "${{ matrix.version }}" = "3.2" ]; then
+            FEATURES="minimal,eddsa"
+          elif [ "${{ matrix.version }}" = "3.5" ]; then
+            FEATURES="minimal,eddsa,mldsa"
+          fi
+
+          if [ "${{ matrix.build }}" = "false" ]; then
+            FEATURES="$FEATURES,dynamic"
+          fi
+
+          cargo build -vv $OPTS --features "$FEATURES"
+          cargo test -vv $OPTS --features "$FEATURES"
+
+      - uses: actions/upload-artifact@v4
+        if: failure()
+        with:
+          name: Build logs OpenSSL version ${{ matrix.name }}
+          path: |
+            target/debug/build/*/output
+

--- a/.github/workflows/pkcs11-provider.yml
+++ b/.github/workflows/pkcs11-provider.yml
@@ -71,7 +71,6 @@ jobs:
           cd pkcs11-provider
           git config --global --add safe.directory \
               /__w/pkcs11-provider/pkcs11-provider
-          git submodule update --init
           meson setup builddir
           meson compile -C builddir
           meson test --num-processes 1 -C builddir

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -91,7 +91,8 @@ nssdb = ["dep:rusqlite", "aes", "hmac", "pbkdf2"]
 
 default = ["standard", "dynamic"]
 
-ecc_all = ["ecdsa", "ec_montgomery", "eddsa", "ecdh"]
+ecc_min = ["ecdsa", "ecdh"]
+ecc_all = ["ecc_min", "ec_montgomery", "eddsa"]
 hash_all = ["hash", "hmac"]
 kdf_all = ["hkdf", "pbkdf2", "sp800_108", "sshkdf", "tlskdf"]
 
@@ -110,3 +111,5 @@ mlkem = ["pkcs11_3_2"]
 mldsa = ["pkcs11_3_2", "hash"]
 
 pqc = ["mlkem", "mldsa"]
+
+minimal = ["sqlitedb", "aes", "ecc_min", "hash_all", "rsa"]

--- a/build.rs
+++ b/build.rs
@@ -2,6 +2,7 @@
 // See LICENSE.txt file for terms
 
 use std::env;
+use std::panic::set_hook;
 use std::path::{Path, PathBuf};
 
 #[derive(Debug)]
@@ -17,6 +18,15 @@ impl Features {
     fn to_bools() -> Features {
         #[cfg(all(feature = "dynamic", feature = "fips"))]
         compile_error!("features `dynamic` and `fips` are mutually exclusive");
+
+        #[cfg(all(
+            feature = "ecdh",
+            not(any(feature = "ecdsa", feature = "ec_montgomery"))
+        ))]
+        compile_error!(
+            "Feature 'ecdh' requires either 'ecdsa' or 'ec_montgomery'"
+        );
+
         Features {
             #[cfg(feature = "fips")]
             fips: true,
@@ -61,6 +71,42 @@ impl bindgen::callbacks::ParseCallbacks for Pkcs11Callbacks {
     }
 }
 
+#[derive(Debug)]
+pub struct OsslCallbacks;
+const OPENSSL_3_0_7: i64 = 0x30000070;
+const OPENSSL_3_2_0: i64 = 0x30200000;
+const OPENSSL_3_5_0: i64 = 0x30500000;
+
+impl bindgen::callbacks::ParseCallbacks for OsslCallbacks {
+    fn int_macro(
+        &self,
+        name: &str,
+        value: i64,
+    ) -> Option<bindgen::callbacks::IntKind> {
+        if name == "OPENSSL_VERSION_NUMBER" {
+            if value < OPENSSL_3_5_0 {
+                #[cfg(any(
+                    feature = "mlkem",
+                    feature = "mldsa",
+                    feature = "fips"
+                ))]
+                panic!("OpenSSL 3.5.0 or later is required for mlkem, mldsa or fips");
+            }
+            if value < OPENSSL_3_2_0 {
+                #[cfg(feature = "eddsa")]
+                panic!("OpenSSL 3.2.0 or later is required for eddsa");
+            }
+            if value < OPENSSL_3_0_7 {
+                panic!(
+                    "OpenSSL 3.0.7 is the minimum viable version. Found {:x}",
+                    value
+                );
+            }
+        }
+        None
+    }
+}
+
 fn ossl_bindings(args: &[&str], out_file: &Path) {
     bindgen::Builder::default()
         .header("ossl.h")
@@ -77,6 +123,7 @@ fn ossl_bindings(args: &[&str], out_file: &Path) {
         .allowlist_item("evp_.*")
         .allowlist_item("BN_.*")
         .allowlist_item("LN_aes.*")
+        .parse_callbacks(Box::new(OsslCallbacks))
         .generate()
         .expect("Unable to generate bindings")
         .write_to_file(out_file)
@@ -84,7 +131,8 @@ fn ossl_bindings(args: &[&str], out_file: &Path) {
 }
 
 fn build_ossl(features: &Features, out_file: &Path) {
-    let sources = std::env::var("KRYOPTIC_OPENSSL_SOURCES").unwrap();
+    let sources = std::env::var("KRYOPTIC_OPENSSL_SOURCES")
+        .expect("Env var KRYOPTIC_OPENSSL_SOURCES is not defined");
     let openssl_path = std::path::PathBuf::from(sources)
         .canonicalize()
         .expect("cannot canonicalize OpenSSL path");
@@ -206,7 +254,33 @@ fn use_system_ossl(out_file: &Path) {
     ossl_bindings(&["-std=c90"], out_file);
 }
 
+fn set_pretty_panic() {
+    set_hook(Box::new(|panic_info| {
+        if let Some(s) = panic_info.payload().downcast_ref::<&str>() {
+            if s != &"panic in a function that cannot unwind" {
+                println!("Compile Error: {s:?}");
+            }
+        } else if let Some(s) = panic_info.payload().downcast_ref::<String>() {
+            if s != "panic in a function that cannot unwind" {
+                println!("Compile Error: {s:?}");
+            }
+        } else {
+            if let Some(location) = panic_info.location() {
+                println!(
+                    "Unrecognized compile error in file '{}' at line {}",
+                    location.file(),
+                    location.line(),
+                );
+            } else {
+                println!("Unknown panic with no location information...");
+            }
+        }
+    }));
+}
+
 fn main() {
+    set_pretty_panic();
+
     let features = Features::to_bools();
 
     let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());

--- a/src/enabled.rs
+++ b/src/enabled.rs
@@ -1,15 +1,6 @@
 // Copyright 2024 Simo Sorce
 // See LICENSE.txt file for terms
 
-#[cfg(all(feature = "dynamic", feature = "fips"))]
-compile_error!("Feature 'dynamic' and 'fips' are mutually exclusive and cannot be enabled together");
-
-#[cfg(all(
-    feature = "ecdh",
-    not(any(feature = "ecdsa", feature = "ec_montgomery"))
-))]
-compile_error!("Feature 'ecdh' requires either 'ecdsa' or 'ec_montgomery'");
-
 #[cfg(feature = "aes")]
 mod aes;
 

